refactor: moved gather model part utility to mpi core

msandre opened the pull request:
GatherModelPartUtility is moved to new mpi core following #2945.

@jcotela should the test in TrilinosApplication/test_examples also be moved? It is designed for 2 MPI processes and doesn't make sense if MPI isn't compiled.

Commit:
refactor: moved gather model part utility to mpi core

GatherModelPartUtility is moved to new mpi core following #2945.

philbucher commented:
Thanks @msandre !

It would be nice if you could refactor the test-example into a test in the MPI-Core
Furthermore (if you have time) you could separate the file into header/source and maybe even clean it a bit. If not I will probably do it at some point

Commit:
test: created test for gather model part utility

The previous test example in TrilinosApplication was converted to a unit test
case and move to the mpi core. The test runs for any number of processes.

msandre commented:
I changed the existing test to work without trilinos and moved it to kratos/mpi. Now it automatically tests instead of printing and it works for any number of procs.

philbucher reviewed (dismissed):
nice, thanks!

If you want you can wait for the opinion of @jcotela , otherwise please go ahead

philbucher committed:
Merge branch 'master' into refactor/move-gather-modelpart-util

RiccardoRossi commented:
if you dont have much hurry i thi k jorfi should take a look. he had well
formed oponions about this

On Fri, Aug 9, 2019, 12:24 PM Philipp Bucher <notifications@github.com>
wrote:

> *@philbucher* approved this pull request.
>
> nice, thanks!
>
> If you want you can wait for the opinion of @jcotela
> <https://github.com/jcotela> , otherwise please go ahead
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5355?email_source=notifications&email_token=AB5PWEP6YOIW7VTVD5OZU6TQDVAUNA5CNFSM4IJJDOMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBDJEVI#pullrequestreview-273060437>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB5PWEKVQHHULDD33JGZ6IDQDVAUNANCNFSM4IJJDOMA>
> .
>

philbucher commented:
> if you dont have much hurry i thi k jorfi should take a look. he had well formed oponions about this
> […](#)
> On Fri, Aug 9, 2019, 12:24 PM Philipp Bucher ***@***.***> wrote: ***@***.**** approved this pull request. nice, thanks! If you want you can wait for the opinion of @jcotela <https://github.com/jcotela> , otherwise please go ahead — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <#5355?email_source=notifications&email_token=AB5PWEP6YOIW7VTVD5OZU6TQDVAUNA5CNFSM4IJJDOMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBDJEVI#pullrequestreview-273060437>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB5PWEKVQHHULDD33JGZ6IDQDVAUNANCNFSM4IJJDOMA> .

I discussed it with him in person. 
Either way, no rush here, then I suggest to wait for his approval

RiccardoRossi commented:
if jordi is ok, ok also frok my side

jcotela commented on kratos/mpi/utilities/gather_modelpart_utility.h:
```diff
@@ -0,0 +1,71 @@
+  void GatherOnMaster();
+
+  template <class TDataType>
+  void GatherOnMaster(Variable<TDataType>& ThisVariable);
```
I would suggest using `const Variable<TDataType>&` as argument type here and in `ScatterFromMaster`

jcotela commented on kratos/mpi/utilities/gather_modelpart_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+{
+    KRATOS_TRY;
+
+    const int mpi_rank = DataCommunicator::GetDefault().Rank();
```
Please use the DataCommunicator of the origin_model_part:
`const DataCommunicator& r_comm = origin_model_part.GetCommunicator().GetDataCommunicator();`
`const int mpi_rank = r_comm.Rank();`

and so on where the DataCommunicator is used.

jcotela commented on kratos/mpi/utilities/gather_modelpart_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+    destination_model_part.Nodes().Unique();
+    if (temp != int(destination_model_part.Nodes().size()))
+        KRATOS_THROW_ERROR(std::logic_error,
+                           "the destination_model_part has repeated nodes", "");
```
Probably a good idea to replace with a `KRATOS_ERROR_IF`

jcotela commented on kratos/mpi/utilities/gather_modelpart_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+      VariablesList* pVariablesList =
+          &destination_model_part.GetNodalSolutionStepVariablesList();
+      destination_model_part.SetCommunicator(
+          Communicator::Pointer(new MPICommunicator(pVariablesList)));
```
Use the origin DataCommunicator also for the destination_model_part:
`Communicator::Pointer(new MPICommunicator(pVariablesList, r_comm)));`

jcotela reviewed (changes requested):
Thanks a lot for taking the lead on this! The test is especially appreciated. I have one request and a few minor comments on the side.

One of the motivations for this mpi communicator refactoring is to be able to support communication using things other than `MPI_COMM_WORLD` in the future. The actual `MPI_Comm` is wrapped by the `DataCommunicator` class, and the `MPICommunicator` can be built using a custom `DataCommunicator` representing, for example, a subset of the processes (think FSI where some processes are "fluid" and some "structure"). For this reason, I would suggest using the `DataCommunicator` instance that comes with the `origin_model_part` instead of the default one. My review comments go mostly in this direction. This will make this utility future-proof when the support for custom communicators is fully operational.

Commit:
refactor: use data communicator from origin model part

Commit:
refactor: use newer error macro

Commit:
fix: add missing const qualifiers to variable arg

Commit:
Merge branch 'refactor/move-gather-modelpart-util' of https://github.com/KratosMultiphysics/Kratos into refactor/move-gather-modelpart-util

msandre commented:
@jcotela done.

jcotela reviewed (dismissed):
:+1: thanks again!

philbucher commented:
I suggest to take a look at codacy, I think there are some relevant recommendations

Commit:
fix: minor codacy suggestions

msandre commented:
> I suggest to take a look at codacy, I think there are some relevant recommendations

done.

jcotela reviewed (approved):
Re-approving

philbucher reviewed (approved):
:+1:
